api/logs/v1/rules_labels_enforcer.go

```diff
@@ -0,0 +1,410 @@
+	for _, name := range l {
+		value := queryParams.Get(name)
+		if value != "" {
+			matchers[name] = value
+		}
+	}
+
+	return matchers
+}
+
+func filterRules(body []byte, contentType string, matchers map[string]string) ([]byte, error) {
+	switch contentType {
+	case contentTypeApplicationJSON:
+		var res prometheusRulesResponse
+		err := json.Unmarshal(body, &res)
+		if err != nil {
+			return nil, err
+		}
+
+		return json.Marshal(filterPrometheusResponse(res, matchers))
+
+	case contentTypeApplicationYAML:
+		var res lokiRulesResponse
+		if err := yaml.Unmarshal(body, &res); err != nil {
+			return nil, err
+		}
+
+		return yaml.Marshal(filterLokiRules(res, matchers))
+
+	default:
+		return nil, errUnknownRulesContentType
+	}
+}
+
+func filterPrometheusResponse(res prometheusRulesResponse, matchers map[string]string) prometheusRulesResponse {
+	if len(matchers) == 0 {
+		res.Data = rulesData{}
+		return res
+	}
+
+	if len(res.Data.RuleGroups) > 0 {
+		filtered := filterPrometheusRuleGroups(res.Data.RuleGroups, matchers)
+		res.Data = rulesData{RuleGroups: filtered}
+	}
+
+	if len(res.Data.Alerts) > 0 {
+		filtered := filterPrometheusAlerts(res.Data.Alerts, matchers)
+		res.Data = rulesData{Alerts: filtered}
+	}
+
+	return res
+}
+
+func filterPrometheusRuleGroups(groups []*ruleGroup, matchers map[string]string) []*ruleGroup {
+	var filtered []*ruleGroup
+
+	for _, group := range groups {
+		var filteredRules []rule
+
+	rules:
+		for _, rule := range group.Rules {
+			for key, value := range matchers {
+				if !rule.Labels().Has(key) || rule.Labels().Get(key) != value {
+					continue rules
+				}
+			}
+
+			filteredRules = append(filteredRules, rule)
+		}
+
+		if len(filteredRules) > 0 {
+			group.Rules = filteredRules
+			filtered = append(filtered, group)
+		}
+	}
+
+	return filtered
+}
+
+func filterPrometheusAlerts(alerts []*alert, matchers map[string]string) []*alert {
+	var filtered []*alert
+
+alerts:
+	for _, alert := range alerts {
+		for key, value := range matchers {
+			if !alert.Labels.Has(key) || alert.Labels.Get(key) != value {
+				continue alerts
+			}
+		}
+
+		filtered = append(filtered, alert)
+	}
+
+	return filtered
+}
+
+func filterLokiRules(res lokiRulesResponse, matchers map[string]string) lokiRulesResponse {
+	if len(matchers) == 0 {
+		return nil
+	}
+
+	filtered := lokiRulesResponse{}
+
+	for name, groups := range res {
+		var filteredGroups []lokiRuleGroup
+
+		for _, group := range groups {
+			var filteredRules []lokiRule
+
+		rules:
+			for _, rule := range group.Rules {
+				for key, value := range matchers {
+					val, ok := rule.Labels[key]
+					if !ok || val != value {
+						continue rules
+					}
+				}
+
+				filteredRules = append(filteredRules, rule)
+			}
+
+			if len(filteredRules) > 0 {
+				group.Rules = filteredRules
+				filteredGroups = append(filteredGroups, group)
+			}
+		}
+
+		if len(filteredGroups) > 0 {
+			filtered[name] = filteredGroups
+		}
+	}
+
+	return filtered
+}
```

Comment:
Following the suggestion above:
```suggestion

func filterLokiRules(res lokiRulesResponse, matchers map[string]string) lokiRulesResponse {
	if len(matchers) == 0 {
		return nil
	}

	filtered := lokiRulesResponse{}

	for name, groups := range res {
		var filteredGroups []lokiRuleGroup

		for _, group := range groups {
			var filteredRules []lokiRule
			for _, rule := range group.Rules {
				if ruleLabelsMatcheMatchers(rule, matchers) {
					filteredRules = append(filteredRules, rule)
				}
			}

			if len(filteredRules) > 0 {
				group.Rules = filteredRules
				filteredGroups = append(filteredGroups, group)
			}
		}

		if len(filteredGroups) > 0 {
			filtered[name] = filteredGroups
		}
	}

	return filtered
}
```